nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java

```diff
@@ -0,0 +1,585 @@
+    }
+
+    /**
+     * Creates a classloader to be used by the selected script engine and the provided script file. This
+     * classloader has the ExecuteScript's classloader as a parent (versus the current thread's context
+     * classloader) and also adds the specified module directory to the classpath. This enables scripts
+     * to use other scripts, modules, etc. without having to build them into the ExecuteScript NAR.
+     * If the parameter is null or empty, ExecuteScript's classloader is returned
+     *
+     * @param modulePath The path to a directory containing modules to be used by the script(s)
+     */
+    private ClassLoader createScriptEngineModuleClassLoader(String modulePath) {
+        URLClassLoader newModuleClassLoader = null;
+        if (StringUtils.isEmpty(modulePath)) {
+            return ExecuteScript.class.getClassLoader();
+        }
+        try {
+            newModuleClassLoader =
+                    new URLClassLoader(
+                            new URL[]{new File(modulePath).toURI().toURL()}, ExecuteScript.class.getClassLoader());
+        } catch (MalformedURLException mue) {
+            getLogger().error("Couldn't find modules directory at " + modulePath, mue);
+        }
+        return newModuleClassLoader;
+    }
+
+    /**
+     * Evaluates the given script body (or file) using the current session, context, and flowfile. The script
+     * evaluation expects a FlowFile to be returned, in which case it will route the FlowFile to success. If a script
+     * error occurs, the original FlowFile will be routed to failure. If the script succeeds but does not return a
+     * FlowFile, the original FlowFile will be routed to no-flowfile
+     *
+     * @param context the current process context
+     * @param session the current process session
+     * @throws ProcessException if any error occurs during script evaluation
+     */
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        synchronized (isInitialized) {
+            if (!isInitialized.get()) {
+                createResources();
+            }
+        }
+        ProcessorLog log = getLogger();
+
+        Bindings bindings = scriptEngine.createBindings();
+        bindings.put("session", session);
+        bindings.put("log", log);
+        bindings.put("REL_SUCCESS", REL_SUCCESS);
+        bindings.put("REL_FAILURE", REL_FAILURE);
+
+        // Find the user-added properties and set them on the script
+        for (Map.Entry<PropertyDescriptor, String> property : context.getProperties().entrySet()) {
+            if (property.getKey().isDynamic()) {
+                // Add the dynamic property and its value as a binding to the script engine
+                if (property.getValue() != null) {
+                    bindings.put(property.getKey().getName(),
+                            context.getProperty(property.getKey()).evaluateAttributeExpressions().getValue());
```

Comment:
This is a little bit weird because I would expect to be able to evaluate FlowFIle Attributes at this point - but we cannot really at this point since there is no FlowFile to evaluate the expression against. Does it make sense to provide the PropertyValue object to the script instead of the String value instead? That way, scripts could evaluate the expressions themselves against a FlowFile?